language/move-symbol-pool/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A global, uniqued cache of strings that is never purged.
+//!
+//! This module provides storage for strings that are meant to remain in use for
+//! the entire running duration of a program. Strings that are stored in this
+//! global, static cache are never evicted, and so the memory consumed by them
+//! can only ever grow.
+//!
+//! The strings can be accessed via the [`Symbol`] type, which acts like a
+//! pointer to the underlying cached string value.
+//!
+//! NOTE: If you're looking for a `#[forbid(unsafe_code)]` attribute here, you
+//! won't find one: symbol-pool (and its inspiration, servo/string-cache) uses
+//! `unsafe` Rust in order to store and dereference `Symbol` pointers to
+//! strings.
+//!
+//! [`Symbol`]: crate::move_symbol_pool::Symbol
+
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::{cmp::Ordering, collections::HashSet, fmt, num::NonZeroU64, sync::Mutex};
+
+type SymbolData = Box<str>;
+static SYMBOL_POOL: Lazy<Mutex<HashSet<SymbolData>>> = Lazy::new(|| Mutex::new(HashSet::new()));
+
+/// Represents a string that has been cached.
```

Comment:
Would you mind writing all of my documentation? lol 
This is glorious!

Reply:
D'aw 😊 I was trying to learn about how docblocks work in Rust and had a lot of fun with it.

TBH I'm not sure all of this works right, I sort of copied the syntax I saw elsewhere for function/type references, and I don't think that syntax is validated by our tooling at the moment -- I'm definitely NOT suggesting it ought to be, but I assume these links are supposed to be used when HTML docs are generated, and who knows if that HTML will work with what I have here 😅

Reply:
I think there is a way to test this so you can see it, let me dig around

And I think any thing in quote blocks is validated? Or maybe we turned it off because it was too annoying? 
I know for a fact I had to turn it off here: https://github.com/diem/diem/blob/main/language/move-lang/src/shared/ast_debug.rs#L13

Reply:
Yeah, I think the blocks are validated and tested -- at least, `cargo xtest -p move-symbol-pool` builds and runs them.